[MSE in Worker] ASSERTION FAILED: The atom string comes from an other thread!

jyavenard opened the pull request:
#### dae8f69439650642c9a2b65f2bb5cebd981265ad
<pre>
[MSE in Worker] ASSERTION FAILED: The atom string comes from an other thread!
<a href="https://bugs.webkit.org/show_bug.cgi?id=287389">https://bugs.webkit.org/show_bug.cgi?id=287389</a>
<a href="https://rdar.apple.com/144502119">rdar://144502119</a>

Reviewed by Eric Carlson and Chris Dumez.

When a MediaSource is used in a DedicatedWorker, the {Audio|Video|Text}TrackPrivateRemote (owned by the MediaPlayerPrivateRemote)
is shared with both the HTMLMediaElement&apos;s tracks on the main thread, and the SourceBuffer&apos;s tracks in the DedicatedWorker.
When the {Audio|Video|Text}Track were constructed, we passed the language and label strings
which constructed the AudioTrack&apos;s label and language AtomStrings from that same string.
When the AtomString is constructed, it mutates the original String&apos;s m_hashAndFlags to indicate
that the string is now part of the AtomString table.
As we then attempted to create another AtomString with that same original String on another thread,
we would get the assertion.
Additionally, we could have used the TrackPrivateRemote&apos;s language and label Strings on multiple threads without
a call to String.isolatedCopy() in between which isn&apos;t allowed (Strings are only ever copied
via a shallow copy).
We now only return an isolated string whenever the TrackPrivateRemote getters are called.

We also extend the same mechanism to InbandTextTrackPrivate&apos;s inBandMetadataTrackDispatchType string.

Fly-by: Have MediaSourcePrivate::currentTime() return zeroTime() instead of invalidTime() when detached (this got missed in 284322@main)

Added tests checking that we can correctly append new init and media segments that contains language information
to a SourceBuffer constructed on a DedicatedWorker.
* LayoutTests/platform/glib/media/media-source/worker/media-managedmse-tracks-worker-expected.txt: Added.
* LayoutTests/media/media-source/worker/media-managedmse-tracks-worker-expected.txt: Added.
* LayoutTests/media/media-source/worker/media-managedmse-tracks-worker.html: Copied from LayoutTests/media/media-source/worker/media-managedmse-worker.html.
* LayoutTests/media/media-source/worker/media-managedmse-worker-expected.txt:
* LayoutTests/media/media-source/worker/media-managedmse-worker.html:
* LayoutTests/media/media-source/worker/worker.js:
(waitForEvent):
(onmessage.async msg):
(logToMain): Deleted.
(onmessage): Deleted.
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/platform/mac-wk1/media/media-source/worker/media-managedmse-tracks-worker-expected.txt: Added.
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::currentTime const):
* Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.cpp:
(WebKit::AudioTrackPrivateRemote::updateConfiguration):
* Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.cpp:
(WebKit::TextTrackPrivateRemote::updateConfiguration):
* Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.cpp:
(WebKit::VideoTrackPrivateRemote::updateConfiguration):
* Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/290263@main">https://commits.webkit.org/290263@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb9d7b9b194e9faa56d8c33cd1b765d5cd9f3a50

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89300 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8824 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44136 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94286 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40061 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91351 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9212 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17187 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68805 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26473 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92302 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7068 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81049 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49166 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6814 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35434 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39168 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77170 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36419 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96115 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16480 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12196 "Found 6 new test failures: fast/css/color-computed-contrast-color-function.html fast/css/color-valid-contrast-color-function.html fast/forms/color/color-input-swatch.html fast/forms/color/input-appearance-color.html imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html imported/w3c/web-platform-tests/screen-orientation/unlock.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77679 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16736 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76837 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76978 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19027 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21394 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19997 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9615 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16494 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21805 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16235 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19686 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18016 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (approved):
LGTM but please ask a media reviewer to take a look too.

eric-carlson reviewed (approved):
Looks good to me as well, thanks

jyavenard committed:
[MSE in Worker] ASSERTION FAILED: The atom string comes from an other thread!
https://bugs.webkit.org/show_bug.cgi?id=287389
rdar://144502119

Reviewed by Eric Carlson and Chris Dumez.

When a MediaSource is used in a DedicatedWorker, the {Audio|Video|Text}TrackPrivateRemote (owned by the MediaPlayerPrivateRemote)
is shared with both the HTMLMediaElement's tracks on the main thread, and the SourceBuffer's tracks in the DedicatedWorker.
When the {Audio|Video|Text}Track were constructed, we passed the language and label strings
which constructed the AudioTrack's label and language AtomStrings from that same string.
When the AtomString is constructed, it mutates the original String's m_hashAndFlags to indicate
that the string is now part of the AtomString table.
As we then attempted to create another AtomString with that same original String on another thread,
we would get the assertion.
Additionally, we could have used the TrackPrivateRemote's language and label Strings on multiple threads without
a call to String.isolatedCopy() in between which isn't allowed (Strings are only ever copied
via a shallow copy).
We now only return an isolated string whenever the TrackPrivateRemote getters are called.

We also extend the same mechanism to InbandTextTrackPrivate's inBandMetadataTrackDispatchType string.

Fly-by: Have MediaSourcePrivate::currentTime() return zeroTime() instead of invalidTime() when detached (this got missed in 284322@main)

Added tests checking that we can correctly append new init and media segments that contains language information
to a SourceBuffer constructed on a DedicatedWorker.
* LayoutTests/platform/glib/media/media-source/worker/media-managedmse-tracks-worker-expected.txt: Added.
* LayoutTests/media/media-source/worker/media-managedmse-tracks-worker-expected.txt: Added.
* LayoutTests/media/media-source/worker/media-managedmse-tracks-worker.html: Copied from LayoutTests/media/media-source/worker/media-managedmse-worker.html.
* LayoutTests/media/media-source/worker/media-managedmse-worker-expected.txt:
* LayoutTests/media/media-source/worker/media-managedmse-worker.html:
* LayoutTests/media/media-source/worker/worker.js:
(waitForEvent):
(onmessage.async msg):
(logToMain): Deleted.
(onmessage): Deleted.
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/platform/mac-wk1/media/media-source/worker/media-managedmse-tracks-worker-expected.txt: Added.
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::currentTime const):
* Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.cpp:
(WebKit::AudioTrackPrivateRemote::updateConfiguration):
* Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.cpp:
(WebKit::TextTrackPrivateRemote::updateConfiguration):
* Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.cpp:
(WebKit::VideoTrackPrivateRemote::updateConfiguration):
* Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.h:

Canonical link: https://commits.webkit.org/290263@main
